common/CrashHandler.cpp

```diff
@@ -0,0 +1,227 @@
+{
+}
+
+CrashHandlerStackWalker::~CrashHandlerStackWalker()
+{
+	if (m_out_file)
+		CloseHandle(m_out_file);
+}
+
+void CrashHandlerStackWalker::OnOutput(LPCSTR szText)
+{
+	if (m_out_file)
+	{
+		DWORD written;
+		WriteFile(m_out_file, szText, static_cast<DWORD>(std::strlen(szText)), &written, nullptr);
+	}
+
+	OutputDebugStringA(szText);
+}
+
+static bool WriteMinidump(HMODULE hDbgHelp, HANDLE hFile, HANDLE hProcess, DWORD process_id, DWORD thread_id,
+	PEXCEPTION_POINTERS exception, MINIDUMP_TYPE type)
+{
+	using PFNMINIDUMPWRITEDUMP =
+		BOOL(WINAPI*)(HANDLE hProcess, DWORD ProcessId, HANDLE hFile, MINIDUMP_TYPE DumpType,
+			PMINIDUMP_EXCEPTION_INFORMATION ExceptionParam, PMINIDUMP_USER_STREAM_INFORMATION UserStreamParam,
+			PMINIDUMP_CALLBACK_INFORMATION CallbackParam);
+
+	PFNMINIDUMPWRITEDUMP minidump_write_dump = hDbgHelp ?
+												   reinterpret_cast<PFNMINIDUMPWRITEDUMP>(GetProcAddress(hDbgHelp, "MiniDumpWriteDump")) :
+                                                   nullptr;
+	if (!minidump_write_dump)
+		return false;
+
+	MINIDUMP_EXCEPTION_INFORMATION mei;
+	PMINIDUMP_EXCEPTION_INFORMATION mei_ptr = nullptr;
+	if (exception)
+	{
+		mei.ThreadId = thread_id;
+		mei.ExceptionPointers = exception;
+		mei.ClientPointers = FALSE;
+		mei_ptr = &mei;
+	}
+
+	return minidump_write_dump(hProcess, process_id, hFile, type, mei_ptr, nullptr, nullptr);
+}
+
+static std::wstring s_write_directory;
+static HMODULE s_dbghelp_module = nullptr;
+static PVOID s_veh_handle = nullptr;
+static bool s_in_crash_handler = false;
+
+static void GenerateCrashFilename(wchar_t* buf, size_t len, const wchar_t* prefix, const wchar_t* extension)
+{
+	SYSTEMTIME st = {};
+	GetLocalTime(&st);
+
+	_snwprintf_s(buf, len, _TRUNCATE, L"%s%scrash-%04u-%02u-%02u-%02u-%02u-%02u-%03u.%s",
+		prefix ? prefix : L"", prefix ? L"\\" : L"",
+		st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond, st.wMilliseconds, extension);
+}
+
+static LONG NTAPI ExceptionHandler(PEXCEPTION_POINTERS exi)
+{
+	if (s_in_crash_handler)
+		return EXCEPTION_CONTINUE_SEARCH;
+
+	switch (exi->ExceptionRecord->ExceptionCode)
+	{
+		case EXCEPTION_ACCESS_VIOLATION:
+		case EXCEPTION_BREAKPOINT:
+		case EXCEPTION_ARRAY_BOUNDS_EXCEEDED:
+		case EXCEPTION_INT_DIVIDE_BY_ZERO:
+		case EXCEPTION_INT_OVERFLOW:
+		case EXCEPTION_PRIV_INSTRUCTION:
+		case EXCEPTION_ILLEGAL_INSTRUCTION:
+		case EXCEPTION_NONCONTINUABLE_EXCEPTION:
+		case EXCEPTION_STACK_OVERFLOW:
+		case EXCEPTION_GUARD_PAGE:
+			break;
+
+		default:
+			return EXCEPTION_CONTINUE_SEARCH;
+	}
+
+	// if the debugger is attached, let it take care of it.
+	if (IsDebuggerPresent())
+		return EXCEPTION_CONTINUE_SEARCH;
+
+	s_in_crash_handler = true;
+
+	wchar_t filename[1024] = {};
+	GenerateCrashFilename(filename, std::size(filename), s_write_directory.empty() ? nullptr : s_write_directory.c_str(), L"txt");
+
+	// might fail
+	HANDLE hFile = CreateFileW(filename, GENERIC_WRITE, 0, nullptr, CREATE_ALWAYS, 0, nullptr);
+	if (hFile)
```

Comment:
From the [CreateFileW](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew#return-value) docs:

>If the function fails, the return value is **INVALID_HANDLE_VALUE**.

Which is `(HANDLE)-1` and not a null pointer.

Reply:
Indeed, nice catch.